auth: added user authentication and authorization.

FaiyazMujawar opened the pull request:
Fixes #6 

**User authentication**

I've added user authentication and authorization with JWT. The user can send their credentials through `POST` method on `api/auth/login`. Upon successful validation of user credentials, the user is responded back with a JWT access token in JSON fomat.

The JWT token payload contains the userID and the user department.

**Format of response on successful validation:**
```
{
    "token":<token>
}
```

**Adding new user:**

New User can be added by sending the required user data in `x-www-form-urlencoded` or `JSON` format, using `POST` method on `api/users/add`. 

**Required fields:**
`zprn`         :  String - 7 characters long, containing all numbers
`firstName` :  String 
`lastName` :  String
`password` :  String - atleast 6 characters long
`department` : String - atleast 3 characters long
`role`          :  String

**Updating user password:**

User can update their password by sending the old password and new password on `api/users/update/password` for updating the password and `api/users/update/profile` for updating profile details, along with the valid JWT access token, using `POST` request. The provided old password is matched with the password in the database. Upon successful match, the new password is hashed and stored in the database.

The admin can update user details by `POST` request on `api/users/update/details/<userID>` with valid JWT token.

**Getting currently logged in user:**
The currently logged in user can get his/her info by sending a `GET` request on `api/users` with the valid JWT token.

**Response:**
```
{
    "zprn":<zprn>,
    "firstName":<firstName>,
    "lastName":<lastName>,
    "department":<department>,
}
```

**Deleting user:**

Moderator(s) can delete any user by sending `DELETE` request on `api/users/<userID>` with valid JWT token.

drigio commented on routes/user.js:
```diff
@@ -0,0 +1,187 @@
+    /*
+     * ADD NEW USER
+     */
+    .post(async (req, res) => {
```
why aren't we having a `api/users/add` or maybe `api/users/create` endpoint for this? It will be much more clearer that way?

FaiyazMujawar replied:
Okay

drigio commented on routes/user.js:
```diff
@@ -0,0 +1,187 @@
+     * UPDATE USER
+     * [This route only updates the user password]
+     */
+    .patch(async (req, res) => {
```
Similarly here we need something like `/update` ? Also we may need to pass in the `userid` parameter here, to identify which user's password is being changed.

FaiyazMujawar replied:
As discussed, only thing that can be changed about the user is the password, which is implemented here. Moreover, the userID of the user requesting the password change is stored in the JWT token itself, so no need to pass the userID. The user can update any other details?

drigio replied:
Yes ofc, in the future we will be having the whole details of user updated. How can a system be possible if we cannot update the user details. We will be needing separate endpoint for this.

drigio replied:
Okay if we go by your way, tell me how will a authority update the details of a user? Or even his password? Like the admin.

FaiyazMujawar replied:
Okay, I'll create a route for updating the  user, while keeping the current route as well.
The current route will be used for changing the details, and new route for authority. I propose keeping both routes as I suppose users don't need to save his details (like userID) on their device,other than the JWT token. userID is stored in the token itself.What say?

drigio replied:
You need to have a parameter for userid. We are developing an api and it needs to be as free and flexible as possible. We cannot constrain the client for doing things. We surely will be passing userid everywhere, ofc it's a major key in many operations. Surely userid will always be sent to the client. And we need to keep only one route for atomicity.

FaiyazMujawar replied:
We'll provide a seperate route for the same. How's that?

drigio reviewed (changes requested):
Everything else looks perfect now. Just a couple of comments, fix that and we are good to merge this PR. :)

FaiyazMujawar committed:
auth: added user authentication and authorization.

User authentication

I've added user authentication and authorization with JWT. The user can send their credentials through `POST` method on `api/auth/login`. Upon successful validation of user credentials, the user is responded back with a JWT access token in JSON fomat.

The JWT token payload contains the userID and the user department.

Format of response on successful validation
{
    "token":<token>
}

Adding new user:

New User can be added by sending the required user data in `x-www-form-urlencoded` format using `POST` method on `api/users/add`. 

Required fields:
`zprn`         :  String - 7 characters long, containing all numbers
`firstName` :  String 
`lastName` :  String
`password` :  String - atleast 6 characters long
`department` : String - atleast 3 characters long
`role`          :  String

Updating user password:
User can update their password by sending the old password and new password on `api/users/update`, along with the valid JWT access token, using `POST` request. The provided old password is matched with the password in the database. Upon successful match, the new password is hashed and stored in the database
